packages/lit-task-auto-top-up/src/Classes/TaskHandler.ts

```diff
@@ -29,11 +32,74 @@ export class TaskHandler {
     recipientDetail: RecipientDetail;
   }): Promise<TaskResult> {
     const { recipientAddress } = recipientDetail;
+    const { noUnexpiredTokensTomorrow, unexpiredTokens } = await this.getExistingTokenDetails({
+      recipientAddress,
+    });
 
-    const capacityTokenIdStr = await mintCapacityCreditNFT({ recipientDetail });
-    await transferCapacityTokenNFT({ capacityTokenIdStr, recipientAddress });
+    if (noUnexpiredTokensTomorrow) {
+      const capacityTokenIdStr = await mintCapacityCreditNFT({ recipientDetail });
+      await transferCapacityTokenNFT({ capacityTokenIdStr, recipientAddress });
 
-    return { capacityTokenIdStr, ...recipientDetail };
+      return { capacityTokenIdStr, result: TaskResultEnum.minted, ...recipientDetail };
+    }
+
+    // If we got here, there should be some `unexpiredTokens` to log for clarity later.
+    return {
+      ...recipientDetail,
+      unexpiredTokens,
+      result: TaskResultEnum.skipped,
+    };
+  }
+
+  private async getExistingTokenDetails({ recipientAddress }: { recipientAddress: string }) {
+    const tomorrow = date.addDays(new Date(), 1);
+    const litContracts = await getLitContractsInstance();
+
+    // :sad_panda:, `getTokensByOwnerAddress()` returns <any> :(
+    const existingTokens: {
+      URI: { description: string; image_data: string; name: string };
+      capacity: {
+        expiresAt: { formatted: string; timestamp: number };
+        requestsPerMillisecond: number;
+      };
+      isExpired: boolean;
+      tokenId: number;
+    }[] =
+      await litContracts.rateLimitNftContractUtils.read.getTokensByOwnerAddress(recipientAddress);
+
+    // Only mint a new token if the recipient...
+    // 1. Has no NFTs at all
+    // 2. All unexpired NFTs they have will expire tomorrow
+    // 3. All of their NFTs are already expired
+    const noUnexpiredTokensTomorrow = existingTokens.every((token) => {
+      // NOTE: `every()` on an empty array === true :)
+      const {
+        capacity: {
+          expiresAt: { timestamp },
+        },
+        isExpired,
+      } = token;
+      if (isExpired) {
+        return true;
+      }
+      return date.isSameDay(new Date(timestamp), tomorrow);
```

Comment:
I would also add a check if the timestamp is before tomorrow, or check that timestamp is later than tomorrow. Just for safety

I suppose `expired` is calculated as `timestamp < now`, so this is checking `timestamp < now && timestamp != tomorrow` which leaves a gap that will return `false` if the NFT expires later today (it will still be expired tomorrow but has not done it yet)
And this is likely to happen as NFTs are checked and minted around the same time, when this runs

Reply:
:+1: Amusingly, I had an itch at the back of my brain about this, and I was already writing a test to check that exact case.  Pushed a fix for this case where this method was made pure, and added unit tests that verify it functions as expected with fixture CapacityToken entries that have explicit times.